velox/functions/prestosql/ArrayFunctions.h

```diff
@@ -316,4 +318,40 @@ struct ArraySumFunction {
   }
 };
 
+/// Function Signature: shuffle(x) -> array
+/// Generate a random permutation of the given array x.
+template <typename TExecCtx, typename T>
+struct ArrayShuffleFunction {
+  static constexpr bool is_deterministic = false;
+  VELOX_DEFINE_FUNCTION_TYPES(TExecCtx);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<velox::Array<T>>& out,
+      const arg_type<velox::Array<T>>& inputArray) {
+    int32_t length = inputArray.size();
+    int32_t indices[length];
+    for (int32_t i = 0; i < length; i++) {
+      indices[i] = i;
+    }
+    // Fisher-Yates shuffle
+    // Randomly swap a pair of indices
+    for (int32_t i = length - 1; i >= 0; i--) {
+      auto randomIdx = folly::Random::rand32(i + 1);
+      std::swap(indices[i], indices[randomIdx]);
+    }
+    for (int32_t i = 0; i < length; i++) {
+      const auto& item = inputArray[indices[i]];
+      if (item.has_value()) {
+        if constexpr (std::is_same_v<T, Varchar>) {
+          out.add_item().setNoCopy(*item);
```

Comment:
same comment as in the other PR

Reply:
Like the other PR, l added a TODO to enable zero copy in the future.